carabiner-to-py/CarabinerToPy.py

```diff
@@ -0,0 +1,80 @@
+import socket
+import edn_format
+import time
+
+class LinkListener():
+    """A simple python client to communicate with carabiner (a Abelton Link connector)
+    Carabiner server must be running to use. Requires edn_format [$pip install edn_format]"""
+    def __init__(self, tcp_ip = '127.0.0.1', tcp_port = 17000, buffer_size = 1024):
+        self.tcp_ip = tcp_ip
+        self.tcp_port = tcp_port
+        self.buffer_size = buffer_size
+        self.s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.s.connect((tcp_ip, tcp_port))
+
+    def decode_edn_msg(self, msg):
+        """Decodes a TCP message from Carabiner to python dictionary"""
+        msg = msg.decode()
+        striped_msg = msg[msg.index('{'):]
+        decoded_msg = edn_format.loads(striped_msg)
+
+        # Because the edf_format package does not return normal dam dicts (or string keywords). What dicts.
+        if type(decoded_msg) is edn_format.immutable_dict.ImmutableDict:
+            decoded_dict = {}
+            for key, value in decoded_msg.dict.items():
+                print()
+                decoded_dict[str(key).strip(':')] = value
+            decoded_msg = decoded_dict
+
+        return decoded_msg
+
+    def status(self):
+        """Wrapper for Status"""
+        self.s.send(b'status')
+        return_msg = self.s.recv(self.buffer_size)
+        return self.decode_edn_msg(return_msg)

```

Comment:
That’s exactly correct, whenever some other participant makes a change to the Link session, Link notifies Carabiner, and Carabiner sends a new `status` message to report the change to all connected clients. So in all the functions where you are sending commands to Carabiner, you cannot assume that the next thing you get back will be a response to your command; it might be a `status` message caused by someone else’s activity. (This is especially true if you don’t have the background thread reading responses!)

So the safe, correct, way to do it is to have all reading from Carabiner performed by the background thread. If you are sending a command that you need to process the response for, set up data structures that the background thread will be able to use to finish the work when that response eventually arrives. You will be able to figure out the right thing to do based on the kind of message. `status` informs you of the latest state of the session timeline, `beat-at-time` is a response to your own beat-at-time message, same for `phase-at-time`, and `time-at-beat`. The commands you can send that change the timeline (`bpm`, `request-beat-at-time`, and `force-beat-at-time`) result in you getting a new `status` message caused by the timeline change you requested.

I’m glad it looks like you have found a suitable clock function in Python, and I agree the documentation could be more helpful about this! Ableton didn’t really expect people to be using Link out-of-process like this, but it’s important for some of us.